File: HashSet-705.py
# Time Complexity : O(1)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : I got the error of index out of bound because of wrong way of list declaration. Updated that later and it worked.


# Your code here along with comments explaining your approach 

class MyHashSet(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.bucket = 1000
        self.bucketItem = 1000
        self.hashset = [False for _ in range(self.bucket)]
        
    
    def hashfunc1(self, key):
        return key % self.bucket
    
    def hashfunc2(self, key):
        return key // self.bucketItem
    
    def add(self, key):
        """
        :type key: int
        :rtype: None
        """
        index1 = self.hashfunc1(key)
        index2 = self.hashfunc2(key)
        if self.hashset[index1] == False:
            self.hashset[index1] = [False for _ in range(self.bucketItem)]
            print self.hashset
        self.hashset[index1][index2] = True
        

    def remove(self, key):
        """
        :type key: int
        :rtype: None
        """
        index1 = self.hashfunc1(key)
        index2 = self.hashfunc2(key)
        if self.hashset[index1] == False:
            return
        self.hashset[index1][index2] = False
        

    def contains(self, key):
        """
        Returns true if this set contains the specified element
        :type key: int
        :rtype: bool
        """
        index1 = self.hashfunc1(key)
        index2 = self.hashfunc2(key)
        
        return self.hashset[index1] and self.hashset[index1][index2]

hs = MyHashSet()
hs.add(1)
hs.add(2)
hs.add(3)
hs.remove(2)
print (hs.contains(2))
print (hs.contains(3))

File: hashSet705.java
package nov172020;

import java.util.LinkedList;

// Source Link: https://leetcode.com/problems/design-hashset/
// https://github.com/super30admin/Design-2
// Problem 4

class hashSet705 {
	private Bucket[] bucketArray;
	private int keyRange;

	/** Initialize your data structure here. */
	public hashSet705() {
		this.keyRange = 769;
		this.bucketArray = new Bucket[this.keyRange];
		for (int i = 0; i < this.keyRange; ++i)
			this.bucketArray[i] = new Bucket();
	}

	protected int _hash(int key) {
		return (key % this.keyRange);
	}

	public void add(int key) {
		int bucketIndex = this._hash(key);
		this.bucketArray[bucketIndex].insert(key);
	}

	public void remove(int key) {
		int bucketIndex = this._hash(key);
		this.bucketArray[bucketIndex].delete(key);
	}

	/** Returns true if this set contains the specified element */
	public boolean contains(int key) {
		int bucketIndex = this._hash(key);
		return this.bucketArray[bucketIndex].exists(key);
	}
}

class Bucket {
	private LinkedList<Integer> container;

	public Bucket() {
		container = new LinkedList<Integer>();
	}

	public void insert(Integer key) {
		int index = this.container.indexOf(key);
		if (index == -1) {
			this.container.addFirst(key);
		}
	}

	public void delete(Integer key) {
		this.container.remove(key);
	}

	public boolean exists(Integer key) {
		int index = this.container.indexOf(key);
		return (index != -1);
	}
}
/*
 * =============================Hash Set In Java====
 */
// - Hashset class implements Set Interface, backed by HashMap instance - This
// class permits null elements
// - The class offers constant time performance for basic operation and size assuming the hash function disappears
// - Underlying data structure for HashSet is HashMap - As it implements the set
// interface,
// - Duplicate values are not allowed
// - Objects that you insert in Hashset are not guaranteed to be inserted in
// same order, objects are inserted based on Hash Code
// - Null elements are allowed in Hash Set
// - Hash set also implements - serializable and cloneable interface
